### PR TITLE
API proposal

### PR DESCRIPTION
**Pull request by wvanbergen**
Here's my proposal to make the API more consistent and well defined. I added yardoc comments to make the types more explicit.

@pickle27 for review/discussion.

### CODE REVIEW

File: lib/active_fulfillment/service.rb [hunk 1]
```diff
@@ -25,15 +25,73 @@ def valid_credentials?
...
+    # @example
+    #
+    #     carrier = ActiveFulfillment::Carrier.new(...)
+    #     response = carrier.fulfill(order.id, order.shipping_address, order.line_items)
+    #     assert response.success?
+    #
+    #
+    # @param order_id [String] Unique identifier for the order
+    # @param shipping_address [?] The address to ship the order to
+    # @param line_items [?] The line items to include in the order.
```
Comment (wvanbergen) [inline lib/active_fulfillment/service.rb]: What is the type of this? We should probably specify that, or maybe include a basic LineItem object.

File: lib/active_fulfillment/service.rb [hunk 2]
```diff
@@ -25,15 +25,73 @@ def valid_credentials?
...
+    #
+    # @example
+    #
+    #     carrier = ActiveFulfillment::Carrier.new(...)
+    #     response = carrier.fulfill(order.id, order.shipping_address, order.line_items)
+    #     assert response.success?
+    #
+    #
+    # @param order_id [String] Unique identifier for the order
+    # @param shipping_address [?] The address to ship the order to
```
Comment (wvanbergen) [inline lib/active_fulfillment/service.rb]: Is this a String, Hash, or a composite object? We should document this, and potentially add a basic Address object that people can use or mimic using duck typing.

File: lib/active_fulfillment/service.rb [hunk 3]
```diff
@@ -25,15 +25,73 @@ def valid_credentials?
...
+    #     tracking_data = carrier.fetch_tracking_data([order.id])
+    #     tracking_data['order_id'].tracking_code
+    #     tracking_data['order_id'].tracking_url
+    #     tracking_data['order_id'].company
+    #
+    # @param options [Array<String>] A list of order IDs for which to retrieve tracking options.
+    # @param options [Hash] Service-specific options.
+    # @return [ActiveFulfillment::TrackingDataResponse]
+    # @note This should be implemented by concrete subclasses.
+    def fetch_tracking_data(order_ids, options = {})
```
Comment (wvanbergen) [inline lib/active_fulfillment/service.rb]: do we also want to add `fetch_all_tracking_data(options =  {})` ?

**Comment by wvanbergen** [comment]
Also do we need all of these: `check_test_mode` `test?`, `test_mode`, and `Base.mode == :test`

File: lib/active_fulfillment/service.rb (hunk 2 above)
Comment (pickle27) [inline lib/active_fulfillment/service.rb]: iirc it is a hash

File: lib/active_fulfillment/service.rb (hunk 1 above)
Comment (pickle27) [inline lib/active_fulfillment/service.rb]: hash again I think

Comment (pickle27) [inline lib/active_fulfillment/service.rb]: Here is an example of me using the Api in the simplest way possible currently:
https://github.com/Shopify/shopify-fulfillment-integration/blob/master/lib/models/fulfillment_service.rb#L48

File: lib/active_fulfillment/service.rb (hunk 2 above)
Comment (wvanbergen) [inline lib/active_fulfillment/service.rb]: we should document what keys in the hash we support/require.

File: lib/active_fulfillment/service.rb (hunk 1 above)
Comment (wvanbergen) [inline lib/active_fulfillment/service.rb]: Again, we should document what keys in the hash we support/require.

Comment (pickle27) [inline lib/active_fulfillment/service.rb]: 100% agree, its pretty cryptic as is

File: lib/active_fulfillment/service.rb (hunk 3 above)
Comment (pickle27) [inline lib/active_fulfillment/service.rb]: +1 to disambiguating request for all or specific data. 

I think for some services all would grow unbounded so maybe its best to leave as only being able to request for certain orders

**Comment by pickle27** [comment]
fyi all the tests failed because I required byebug in the test_helper

**Comment by wvanbergen** [comment]
yeah, I removed that locally but failed to commit it.

**Comment by wvanbergen** [comment]
Going to continue to work on this this week.

**Comment by kevinhughes27** [comment]
how is this coming along? I've been on vacay but am back now if you want to pair up this or next week

**Comment by wvanbergen** [comment]
Going on vacation, but I am willing to get back to this in March.

**Comment by kevinhughes27** [comment]
cool ping me when you are back and we'll have another pairing sesh

**Comment by wvanbergen** [comment]
@kevinhughes27 ping. I am back!

**Comment by kevinhughes27** [comment]
lol - should we find a new owner to look at these changes?

**Pull request closed without merging**